src/main_thread/decrypt/content_decryptor.ts

```diff
@@ -692,6 +705,22 @@ export default class ContentDecryptor extends EventEmitter<IContentDecryptorEven
       return false;
     }
 
+    /**
+     * On Safari using Directfile, the old EME implementation triggers
+     * the "webkitneedkey" event instead of "encrypted". There's an issue in Safari
+     * where "webkitneedkey" fires too early before all tracks are added from an HLS playlist.
+     * Safari incorrectly assumes some keys are missing for these tracks,
+     * leading to repeated "webkitneedkey" events. Because RxPlayer recognizes
+     * it already has a session for these keys and ignores the events,
+     * the content remains frozen. To resolve this, the session is re-created.
+     */
+    const forceSessionRecreation =
+      eme.implementation === "webkit" && this._context.isDirectFile;
```

Comment:
Couldn't we just emit that property at event-dispatching?

We could even maybe remove the need for the context.isDirectfile option

Reply:
So I should patch the "webkitneedkey" event ? I need to listen to the event and update the event value before sending it to the listener
We didn't experience the issue with DASH, only with HLS and direct file, this is why I have done this

Reply:
Yes at the webkitneedkey event.

Are we sure that this issue only appears for directfile by the way? E.g. it could be that MediaSource-based playback + webkit EME API repeats the issue.

---

I'm also just wondering right now, isn't there a risk of multiple concurrent `webkitneedkey` events while license-fetching is still under way in which case we will keep closing sessions and making license requests a lot of time?

When do the `webkitneedkey` event stop? At the `generateRequest` step or a the license-pushing step?